source/extensions/clusters/aggregate/cluster_lb.cc

```diff
@@ -0,0 +1,132 @@
+#include "extensions/clusters/aggregate/cluster_lb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+AggregateClusterLoadBalancer::AggregateClusterLoadBalancer(
+    Upstream::ClusterManager& cluster_manager, const std::vector<std::string>& clusters,
+    Upstream::ClusterStats& stats, Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+    const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+    : cluster_manager_(cluster_manager), clusters_(clusters), stats_(stats), runtime_(runtime),
+      random_(random), common_config_(common_config) {}
+
+void AggregateClusterLoadBalancer::initialize() {
+  ASSERT(!initialized_);
+  initialized_ = true;
+  refresh();
+  updatePrioritySetCallbacks(
+      clusters_,
+      [this](uint32_t, const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); },
+      [this](const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+}
+
+void AggregateClusterLoadBalancer::refresh() {
+  std::pair<Upstream::PrioritySetImpl,
+            std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>
+      pair = linearizePrioritySet();
+  if (pair.first.hostSetsPerPriority().empty()) {
+    load_balancer_ = nullptr;
+  } else {
+    load_balancer_ =
+        std::make_unique<AggregateClusterLoadBalancer::LoadBalancerImpl>(*this, std::move(pair));
+  }
+}
+
+void AggregateClusterLoadBalancer::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    if (deleted_clusters_.count(cluster.info()->name())) {
```

Comment:
do we really need to track deleted clusters? dont we only need to worry about this during the `onClusterRemoval` call? once that call completes the cluster should be removed from the cluster manager so no need to track it?

Reply:
We don't need to track deleted clusters. We only to remember which cluster is being deleted. The pro of having deleted cluster is that we don't need to get deleted thread local cluster again and again.

Reply:
But we only need to know that for the lifetime of the `onClusterRemoved` callback? You'll just get nullptr back when you look it up after that?

Reply:
Yes, correct. The cost comparison between getting thread local clusters from cluster manager and getting delete cluster from `set` matters.

Reply:
i wouldnt expect perf to matter much here, its not on the hot path (assuming we get rid of the lazy init) and the cluster lookup should be fairly fast to begin with (just a hash map lookup). i would opt for readability over perf concerns here

Reply:
Removed the `deleted_cluster` and use predicate as a parameter.